src/controllers/legacycontrollersettings.h

```diff
@@ -413,14 +423,158 @@ class LegacyControllerEnumSetting
+    }
+    static inline bool match(const QDomElement& element) {
+        return element.hasAttribute("type") &&
+                QString::compare(element.attribute("type"),
+                        "color",
+                        Qt::CaseInsensitive) == 0;
+    }
+
+  protected:
+    LegacyControllerColorSetting(const QDomElement& element,
+            QColor currentValue,
+            QColor defaultValue)
+            : AbstractLegacyControllerSetting(element),
+              m_defaultValue(defaultValue),
+              m_savedValue(currentValue) {
+    }
+
+    virtual QWidget* buildInputWidget(QWidget* parent) override;
+
+  private:
+    QColor m_defaultValue;
+    QColor m_savedValue;
+
+    QColor m_editedValue;
+
+    FRIEND_TEST(ControllerS4MK3SettingTest, ensureLibrarySettingValueAndEnumEquals);
+    FRIEND_TEST(LegacyControllerMappingSettingsTest, enumSettingEditing);
+};
+
+class LegacyControllerFileSetting
+        : public LegacyControllerSettingFactory<LegacyControllerFileSetting>,
+          public AbstractLegacyControllerSetting {
+  public:
+    LegacyControllerFileSetting(const QDomElement& element);
+
+    virtual ~LegacyControllerFileSetting() = default;
+
+    QJSValue value() const override {
+        return QJSValue(stringify());
+    }
+
+    QString stringify() const override {
+        return m_savedValue.absoluteFilePath();
+    }
+    void parse(const QString& in, bool* ok) override;
+    bool isDefault() const override {
+        return m_savedValue == m_defaultValue;
+    }
+    bool isDirty() const override {
+        return m_savedValue != m_editedValue;
+    }
+
+    virtual void save() override {
+        m_savedValue = m_editedValue;
+    }
+
+    virtual void reset() override {
+        m_editedValue = m_defaultValue;
+        emit valueReset();
+    }
+
+    /// @brief Whether or not this setting definition and its current state are
+    /// valid. Validity scope includes a known default/current/dirty option.
+    /// @return true if valid
+    bool valid() const override {
+        return AbstractLegacyControllerSetting::valid() &&
+                (m_defaultValue == m_savedValue || m_savedValue.exists());
+    }
+
+    static AbstractLegacyControllerSetting* createFrom(const QDomElement& element) {
+        return new LegacyControllerFileSetting(element);
+    }
+    static inline bool match(const QDomElement& element) {
+        return element.hasAttribute("type") &&
+                QString::compare(element.attribute("type"),
+                        "file",
+                        Qt::CaseInsensitive) == 0;
+    }
+
+  protected:
+    LegacyControllerFileSetting(const QDomElement& element,
+            const QFileInfo& currentValue,
+            const QFileInfo& defaultValue)
+            : AbstractLegacyControllerSetting(element),
+              m_defaultValue(defaultValue),
+              m_savedValue(currentValue) {
+    }
+
+    virtual QWidget* buildInputWidget(QWidget* parent) override;
```

Comment:
already implied by `override`. at least my clangd-based linter complains about this.
```suggestion
    QWidget* buildInputWidget(QWidget* parent) override;
```